Refactor Explore query field

davkal opened the pull request:
- extract typeahead field that only contains logic for the typeahead
  mechanics
- added tons of TS types
- <s>renamed</s> QueryField <s>to PromQueryField</s>, a wrapper around TypeaheadField
  that deals with Prometheus-specific concepts
- QueryField creates a promql typeahead by providing the handlers
  for producing suggestions, and for applying suggestions
- The `refresher` promise is needed to trigger a render once an async
  action in the wrapper returns.

This is prep work for a composable query field to be used by Explore, as
well as editors in datasource plugins.

davkal committed:
Refactor Explore query field

- extract typeahead field that only contains logic for the typeahead
  mechanics
- renamed QueryField to PromQueryField, a wrapper around TypeaheadField
  that deals with Prometheus-specific concepts
- PromQueryField creates a promql typeahead by providing the handlers
  for producing suggestions, and for applying suggestions
- The `refresher` promise is needed to trigger a render once an async
  action in the wrapper returns.

This is prep work for a composable query field to be used by Explore, as
well as editors in datasource plugins.

davkal commented:
Kept renaming of the QueryField to a later PR to make diffs easier on this one.

torkelo commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,282 @@
+    }
+  };
+
+  handleTypeahead = (typeahead: TypeaheadInput): TypeaheadOutput => {
```
this looks like a pretty complex & hairy function that looks very well suited for tests

davkal replied:
Extracted the crucial parts and wrote tests.

torkelo commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,282 @@
+        additionalPlugins={this.plugins}
+        cleanText={cleanText}
+        initialValue={this.props.initialQuery}
+        onTypeahead={this.handleTypeahead}
```
I think we should align on the airbnb style guide of naming event / callback functions "onTypeahead" , onSomethingChanged etc, instead of handle prefix.

torkelo commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,282 @@
+        [EMPTY_METRIC]: values,
+      };
+      this.setState({ labelValues });
+    } catch (e) {
```
Is there a way we can centralize try/catch error handling more? Maybe via component decorators?  https://github.com/influxdata/chronograf/blob/master/ui/src/shared/decorators/errors.tsx

davkal replied:
Needs a better approach in general. I'd do this in a separate PR. For this one I harmonized the handling to simply `console.error()`

torkelo reviewed (changes requested):
Looks good but some minor ideas for improvements :)

davkal committed:
Added typeahead handling tests

- extracted context-to-suggestion logic to make it testable
- kept DOM-dependent parts in main onTypeahead funtion

davkal committed:
simplified error handling in explore query field

davkal commented:
PTAL @torkelo 
Couldnt find a straightforward way to test `onTypeahead` as is, so I split it into DOM-dependent and DOM-independent part.

davkal committed:
Refactor query suggestions

- use monaco's suggestion types (roughly), see
  https://github.com/Microsoft/monaco-editor/blob/f6fb545/monaco.d.ts#L4208
- suggest functions and metrics in empty field (ctrl+space)
- copy and expand prometheus function docs from prometheus datasource
  (will be migrated back to the datasource in the future)

alexanderzobnin commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,303 @@
+    this.fetchMetricNames();
+  }
+
+  componentWillReceiveProps(nextProps) {
```
`componentWillReceiveProps()` method is marked as [deprecated](https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops). There're some anti-patterns related to it:
https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#anti-pattern-erasing-state-when-props-change
I think this is not exactly the same case, but if we can avoid using deprecated components it would be great.

alexanderzobnin commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,303 @@
+  return suggestion;
+}
+
+class PromQueryField extends React.Component<any, any> {
```
I think explicit types for props and state would be better than <any, any>, what do you think?

davkal committed:
Added prop and state types, removed unused cwrp

alexanderzobnin commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,310 @@
+  };
+
+  // Keep this DOM-free for testing
+  getTypeahead(
```
I know, determining suggestions is a complex process, but this function looks really hard for reading. Could we break it out into something like 
```js
    if (_.includes(wrapperClasses, 'context-range')) {
      return getRangeSuggestions();
    } else if (_.includes(wrapperClasses, 'context-labels') && metric) {
      return getLabelSuggestions();
    } else if (_.includes(wrapperClasses, 'context-labels') && !metric) {
```

alexanderzobnin commented on public/app/containers/Explore/PromQueryField.tsx:
```diff
@@ -0,0 +1,310 @@
+  };
+
+  async fetchLabelValues(key) {
+    const url = `/api/v1/label/${key}/values`;
```
This part looks similar for next functions:
```js
    const url = `/api/v1/label/${key}/values`;
    try {
      const res = await this.request(url);
      const body = await (res.data || res.json());
```
is it possible to extract it and reuse?

davkal committed:
Split up suggestion processing for code readability
